src/Frontend-GNOME/Views/MessageTextView.cs

```diff
@@ -406,6 +414,53 @@ public void AddMessage(MessageModel msg, bool addLinebreak, bool showTimestamps)
                     } else {
                         buffer.Insert(ref iter, fmsgti.Text);
                     }
+                } else if (msgPart is ImageMessagePartModel) {
+                    var imgpart = (ImageMessagePartModel) msgPart;
+                    Uri uri = null;
+                    try {
+                        uri = new Uri(imgpart.ImageFileName);
+                    } catch (UriFormatException) {
+                        AddAlternativeText(buffer, ref iter, imgpart);
+                    }
+                    switch (uri.Scheme) {
+                        case "smuxi-emoji":
+                            var shortName = uri.Host;
+                            var iconCache = new IconCache("emoji");
+                            var unicode = Emojione.ShortnameToUnicode(shortName);
+                            if (unicode == null) {
+                                AddAlternativeText(buffer, ref iter, imgpart);
+                            } else {
+                                var emojiName = unicode + ".png";
+                                string emojiPath;
+                                if (iconCache.TryGetIcon("emojione", emojiName, out emojiPath)) {
+                                    var emojiFile = new FileInfo(emojiPath);
+                                    if (emojiFile.Exists && emojiFile.Length > 0) {
+                                        var pix = new Gdk.Pixbuf(emojiPath);
+                                        buffer.InsertPixbuf(ref iter, pix);
+                                    }
+                                } else {
+                                    // add a mark here so we know where to insert the pixbuf
+                                    // once we've downloaded the file
+                                    var mark = new Gtk.TextMark(null, true);
+                                    buffer.AddMark(mark, iter);
+                                    var emojiUrl = Emojione.UnicodeToUrl(unicode);
+                                    iconCache.DownloadFile("emojione", emojiName, emojiUrl,
+                                        (path) => {
+                                            var markIter = buffer.GetIterAtMark(mark);
```

Comment:
Memory Corruption: the callback must be wrapped in a UI thread callback, else you will corrupt glib owned memory here. If the method would be named BeginDownloadFile() then this would be more obvious ^^